java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/htmlui/Browser.java

```diff
@@ -18,538 +18,168 @@
+                        else if (lower != ' ' && lower != '\n') state = 0;
+                        break;
+                    case 2: if (lower == 'o') state = 3; else state = 0; break;
+                    case 3: if (lower == 'd') state = 4; else state = 0; break;
+                    case 4: if (lower == 'y') state = 5; else state = 0; break;
+                    case 5: if (lower == '>') state = 500;
+                        else if (lower != ' ' && lower != '\n') state = 0;
+                        break;
+                }
+                sb.append((char)ch);
+                if (state == 500) {
+                    emitScript(sb, id);
+                    state = 1000;
                 }
-                // Show.show(impl, page); TBD
-            };
-            return this;
-        }
-        public Config browser(Consumer<URI> urlOpener) {
-            this.browser = urlOpener;
-            return this;
-        }
-
-        /** The port to start the server at.
-         * By default a random port is selected.
-         * @param port the port
-         * @return this instance
-         */
-        public Config port(int port) {
-            this.port = port;
-            return this;
-        }
-
-        /** Enable or disable debugging. The default value is taken from a property
-         * {@code com.dukescript.presenters.browserDebug}. If the property is
-         * not specified, then the default value is {@code false}.
-         *
-         * @param debug true or false
-         * @return this instance
-         * @since 1.8
-         */
-        public Config debug(boolean debug) {
-            this.debug = debug;
-            return this;
-        }
-
-        final Consumer<URI> getBrowser() {
-            return browser;
-        }
-
-        final int getPort() {
-            if (port != null) {
-                return port;
             }
-            String browserPort = System.getProperty("com.dukescript.presenters.browserPort"); // NOI18N
-            try {
-                return Integer.parseInt(browserPort);
-            } catch (NumberFormatException ex) {
-                return -1;
+            if (state != 1000) {
+                emitScript(sb, id);
             }
         }
+        return sb.toString();
     }
 
-    static <Response> void cors(HttpServer<?, Response, ?, ?> s, Response r) {
-        s.setCharacterEncoding(r, "UTF-8");
-        s.addHeader(r, "Access-Control-Allow-Origin", "*");
-        s.addHeader(r, "Access-Control-Allow-Credentials", "true");
-        s.addHeader(r, "Access-Control-Allow-Headers", "Content-Type");
-        s.addHeader(r, "Access-Control-Allow-Methods", "GET, POST, DELETE, PUT");
+    private void emitScript(StringBuilder sb, String id) throws IOException {
+        sb.append("<script id='exec' type='text/javascript'>\n"
+                + "(function () {\n"
+                + "  const self = this;\n"
+                + "  this.vscode = acquireVsCodeApi();\n"
```

Comment:
The usual recommendation is to "hide" the `vscode` API object in a local variable to avoid its misuse by untrusted code running in the browser. E.g. I'd store it only in `const vscode = ` rather than putting it into `this.vscode`.

Reply:
Fixed.